<launch>
  <node pkg="nodelet" type="nodelet" name="multi_plane_manager"
        args="manager" output="screen"/>
  <node pkg="nodelet" type="nodelet"
        name="xfilter"
        args="load pcl/PassThrough multi_plane_manager">
    <remap from="~input" to="tilt_laser_listener/output_cloud" />
    <rosparam>
      filter_field_name: x
      filter_limit_min: 0.0
      filter_limit_max: 3.0
    </rosparam>
  </node>
  <node pkg="nodelet" type="nodelet"
        name="yfilter"
        args="load pcl/PassThrough multi_plane_manager">
    <remap from="~input" to="xfilter/output" />
    <rosparam>
      filter_field_name: y
      filter_limit_min: -2.0
      filter_limit_max: 2.0
    </rosparam>
  </node>
  <node pkg="nodelet" type="nodelet"
        name="zfilter"
        args="load pcl/PassThrough multi_plane_manager">
    <remap from="~input" to="yfilter/output" />
    <rosparam>
      filter_field_name: z
      filter_limit_min: -2.0
      filter_limit_max: 2.0
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet"
        args="load pcl/VoxelGrid multi_plane_manager"
        name="voxel_grid">
    <remap from="~input" to="zfilter/output"/>
    <rosparam>
      leaf_size: 0.02
      filter_limit_min: -10.0
      filter_limit_max: 10.0
    </rosparam>
  </node>
  
  <node pkg="nodelet" type="nodelet"
        args="load pcl/NormalEstimationOMP multi_plane_manager"
        name="normal_estimation">
    <remap from="~input" to="voxel_grid/output" />
    <rosparam>
      radius_search: 0
      k_search: 40
      # 0, => ANN, 1 => FLANN, 2 => Organized
      spatial_locator: 1
    </rosparam>
  </node>

  
  <node pkg="nodelet" type="nodelet" name="normal_concat"
        args="load jsk_pcl/NormalConcatenater multi_plane_manager"
        clear_params="true">
    <remap from="~input" to="voxel_grid/output"/>
    <remap from="~normal" to="normal_estimation/output"/>
  </node>

  <node pkg="nodelet" type="nodelet"
        args="load jsk_pcl/PlaneConcatenator multi_plane_manager"
        name="plane_concatenator">
    <remap from="~input" to="normal_direction_filtered_cloud/output" />
    <remap from="~input/indices" to="multi_plane_region_growing/output/inliers" />
    <remap from="~input/polygons" to="multi_plane_region_growing/output/polygons" />
    <remap from="~input/coefficients" to="multi_plane_region_growing/output/coefficients" />
  </node>
  
  <node pkg="nodelet" type="nodelet"
        args="load jsk_pcl/RegionGrowingMultiplePlaneSegmentation multi_plane_manager"
        name="multi_plane_region_growing">
    <remap from="~input" to="normal_direction_filtered_cloud/output" />
    <remap from="~input_normal" to="normal_direction_filtered_normal/output" />
    <rosparam>
      cluster_tolerance: 0.3
    </rosparam>
  </node>
  
  <node pkg="nodelet" type="nodelet"
        args="load jsk_pcl/MultiPlaneSACSegmentation multi_plane_manager"
        name="multi_plane_sac_segmentation">
    <remap from="~input" to="voxel_grid/output" />
    <remap from="~input_clusters" to="multi_plane_region_growing/output/inliers" />
    <remap from="~input_normal" to="normal_estimation/output" />
    <rosparam>
      use_imu_parallel: false
      use_imu_perpendicular: false
      use_normal: false
      use_clusters: true
    </rosparam>
  </node>
  <node pkg="nodelet" type="nodelet"
        args="load jsk_pcl/ClusterPointIndicesDecomposer multi_plane_manager"
        name="multi_plane_region_growing_decomposer">
    <remap from="~input" to="normal_direction_filtered_cloud/output" />
    <remap from="~target" to="multi_plane_region_growing/output/inliers" />
    <rosparam>
      publish_tf: false
      publish_clouds: false
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet"
        args="load jsk_pcl/NormalDirectionFilter multi_plane_manager"
        name="normal_direction_filter">
    <remap from="~input" to="normal_estimation/output" />
    <remap from="~input_imu" to="/multisense/imu/imu_data" />
    <rosparam>
      use_imu: true
    </rosparam>
  </node>
  <node pkg="nodelet" type="nodelet"
        args="load pcl/ExtractIndices multi_plane_manager"
        name="normal_direction_filtered_cloud">
    <remap from="~input" to="voxel_grid/output" />
    <remap from="~indices" to="normal_direction_filter/output" />
  </node>
  <node pkg="nodelet" type="nodelet"
        args="load pcl/ExtractIndices multi_plane_manager"
        name="normal_direction_filtered_normal">
    <remap from="~input" to="normal_estimation/output" />
    <remap from="~indices" to="normal_direction_filter/output" />
  </node>

  <node pkg="nodelet" type="nodelet"
        name="plane_extraction"
        args="load jsk_pcl/MultiPlaneExtraction multi_plane_manager"
        output="screen">
    <remap from="~input" to="voxel_grid/output" />
    <remap from="~input_polygons" to="plane_concatenator/output/polygons" />
    <remap from="~input_coefficients" to="plane_concatenator/output/coefficients" />
    <rosparam>
      use_indices: false
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet" name="euclidean_clustering"
        args="load jsk_pcl/EuclideanClustering multi_plane_manager" output="screen">
    <remap from="~input" to="/plane_extraction/output" />
    <rosparam>
      tolerance: 0.02
      min_size: 100
    </rosparam>
  </node>
  <node pkg="nodelet" type="nodelet"
        args="load jsk_pcl/ClusterPointIndicesDecomposer multi_plane_manager"
        name="euclidean_clustering_decomposer">
    <remap from="~input" to="/plane_extraction/output" />
    <remap from="~target" to="euclidean_clustering/output" />
    <rosparam>
      publish_tf: false
      publish_clouds: false
    </rosparam>
  </node>
  
  
</launch>
